Utils/hydro_enforce_inout_solvability.cpp

```diff
@@ -0,0 +1,307 @@
+        const auto islow = ori.isLow();
+        const auto ishigh = ori.isHigh();
+
+        // Multifab for normal velocity
+        auto& vel_mf = vels_vec[lev][dir];
+
+        // mask iMFs for the respective velocity direction
+        auto& inflow_mask = inflow_masks[dir];
+        auto& outflow_mask = outflow_masks[dir];
+
+        IndexType::CellIndex dir_index_type = (vel_mf->ixType()).ixType(dir);
+        // domain extent indices for the velocities
+        int dlo;
+        if (dir_index_type == IndexType::CellIndex::CELL) {
+            // lower boundary is at -1 for cell-centered velocity
+            dlo = domain.smallEnd(dir) - 1;
+        } else {
+            // lower boundary is at  0 for face-centered velocity
+            dlo = domain.smallEnd(dir);
+        }
+        int dhi = domain.bigEnd(dir) + 1;
+
+        // get BCs for the normal velocity and set the boundary index
+        // based on low or high side
+        const BCRec ibcrec = bc_type[dir];
+        int bc, bndry;
+        if (side == Orientation::low) {
+            bc = ibcrec.lo(dir);
+            bndry = dlo;
+        } else {
+            bc = ibcrec.hi(dir);
+            bndry = dhi;
+        }
+
+        // limit influx/outflux calculations to the in-out boundaries only
+        if (bc == BCType::direction_dependent) {
+            for (MFIter mfi(*vel_mf, TilingIfNotGPU()); mfi.isValid(); ++mfi) {
+
+                Box box = mfi.validbox();
+
+                // include ghost cells along normal velocity for cell-centered
+                // not for face-centered as boundary lies in valid region
+                if (dir_index_type == IndexType::CellIndex::CELL) {
+                    box.grow(dir, 1);
+                }
+
+                // include boundary corners if specified
+                if (corners) {
+                    box.grow((dir+1)%AMREX_SPACEDIM, 1);
+                    box.grow((dir+2)%AMREX_SPACEDIM, 1);
+                }
+
+                // Enter further only if the box bndry is at the domain bndry
+                if ((islow && (box.smallEnd(dir) == dlo))
+                 || (ishigh && (box.bigEnd(dir) == dhi))) {
+
+                    // create a 2D box normal to dir at the low/high bndry
+                    Box box2d(box); box2d.setRange(dir, bndry);
+
+                    auto vel_arr = vel_mf->array(mfi);
+                    auto in_mask = inflow_mask.array(mfi);
+                    auto out_mask = outflow_mask.array(mfi);
+
+                    // tag cells as inflow or outflow by checking vel direction
+                    ParallelFor(box2d, [=] AMREX_GPU_DEVICE (int i, int j, int k)
+                    {
+                        if ((side == Orientation::low && vel_arr(i,j,k) >= 0)
+                         || (side == Orientation::high && vel_arr(i,j,k) <= 0)) {
+                            in_mask(i,j,k) = 1;
+                        } else {
+                            out_mask(i,j,k) = 1;
```

Comment:
Since the in_mask is the opposite of out_mask, why is it necessary to have two separate masks?

Reply:
Oh that's a good point! For instance, I could have a value of -1 for inflow and +1 for outflow as these masks are integers and not boolean and then use those values for the if conditions in the reduction. 

Thanks, I'll try this out and let you know when I've implemented it.

Reply:
I have now implemented a single in/out mask that sets to -1 for inflow, +1 for outflow, and 0 for interior cells.

Reply:
https://github.com/AMReX-Fluids/AMReX-Hydro/pull/129/commits/6d567d5a4fdaf5de83c1232a116c648d2e67d1af